test/martha/martha_v3.test.js

```diff
@@ -0,0 +1,154 @@
+    await marthaV3(mockRequest({ body: { 'uri': 'dos://abc/123' } }), response);
+    t.is(response.statusCode, 400);
+});
+
+test.serial('martha_v3 should return 400 if no data is posted with the request', async (t) => {
+    const response = mockResponse();
+    await marthaV3(mockRequest({}), response);
+    t.is(response.statusCode, 400);
+});
+
+test.serial('martha_v3 should return 400 if given a \'url\' with an invalid value', async (t) => {
+    const response = mockResponse();
+    await marthaV3(mockRequest({ body: { url: 'Not a valid URI' } }), response);
+    t.is(response.statusCode, 400);
+});
+
+test.serial('martha_v3 should return 502 if Data Object resolution fails', async (t) => {
+    getJsonFromApiStub.restore();
+    sandbox.stub(apiAdapter, getJsonFromApiMethodName).rejects(new Error('Data Object Resolution forced to fail by testing stub'));
+    const response = mockResponse();
+    await marthaV3(mockRequest({ body: { 'url': 'dos://abc/123' } }), response);
+    const result = response.send.lastCall.args[0];
+    t.true(result instanceof Error);
+    t.is(response.statusCode, 502);
+});
+
+test.serial('martha_v3 should return 502 if key retrieval from bond fails', async (t) => {
+    getJsonFromApiStub.restore();
+    sandbox.stub(apiAdapter, getJsonFromApiMethodName).rejects(new Error('Bond key lookup forced to fail by testing stub'));
+    const response = mockResponse();
+    await marthaV3(mockRequest({ body: { 'url': 'dos://abc/123' } }), response);
+    const result = response.send.lastCall.args[0];
+    t.true(result instanceof Error);
+    t.is(response.statusCode, 502);
+});
+
+test.serial('martha_v3 calls bond Bond with the "dcf-fence" provider when the Data Object URL host is not "dg.4503"', async (t) => {
+    const response = mockResponse();
+    await marthaV3(mockRequest({ body: { 'url': 'dos://abc/123' } }), response);
+    const requestedBondUrl = getJsonFromApiStub.secondCall.args[0];
+    const matches = requestedBondUrl.match(bondRegEx);
+    t.truthy(matches, 'Bond URL called does not match Bond URL regular expression');
+    const expectedProvider = 'dcf-fence';
+    const actualProvider = matches[2];
+    t.is(actualProvider, expectedProvider);
+});
+
+test.serial('martha_v3 calls bond Bond with the "fence" provider when the Data Object URL host is "dg.4503"', async (t) => {
+    const response = mockResponse();
+    await marthaV3(mockRequest({ body: { 'url': 'drs://dg.4503/this_part_can_be_anything' } }), response);
+    const requestedBondUrl = getJsonFromApiStub.secondCall.args[0];
+    const matches = requestedBondUrl.match(bondRegEx);
+    t.truthy(matches, 'Bond URL called does not match Bond URL regular expression');
+    const expectedProvider = 'fence';
+    const actualProvider = matches[2];
+    t.is(actualProvider, expectedProvider);
+});
+
+test.serial('martha_v3 does not call Bond or return SA key when the Data Object URL host endswith ".humancellatlas.org', async (t) => {
```

Comment:
Related to the prior comment, there should be a test that JDR-ish requests do not talk to bond.

Reply:
Yes. I have added that test, but as part of [WA-179](https://broadworkbench.atlassian.net/browse/WA-179).